core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -0,0 +1,189 @@
+ */
+public class ChecksumBlobStoreFormat<T extends ToXContent> extends BlobStoreFormat<T> {
+
+    public static final String TEMP_FILE_PREFIX = "pending-";
+
+    private static final XContentType DEFAULT_X_CONTENT_TYPE = XContentType.JSON;
+
+    // The format version
+    public static final int VERSION = 1;
+
+    private static final int BUFFER_SIZE = 4096;
+
+    protected final XContentType xContentType;
+
+    protected final boolean wrapper;
+
+    protected final boolean compress;
+
+    public final String codec;
+    /**
+     * @param codec codec name
+     * @param blobNameFormat format of the blobname in {@link String#format} format
+     * @param reader        prototype object that can deserialize T from XContent
+     * @param wrapper        true if content should be surrounded with extra start/endObject during serialization
+     * @param compress      true if the content should be compressed
+     * @param xContentType   content type that should be used for write operations
+     */
+    public ChecksumBlobStoreFormat(String codec, String blobNameFormat, FromXContentBuilder<T> reader, ParseFieldMatcher parseFieldMatcher, boolean wrapper, boolean compress, XContentType xContentType) {
+        super(blobNameFormat, reader, parseFieldMatcher);
+        this.xContentType = xContentType;
+        this.wrapper = wrapper;
+        this.compress = compress;
+        this.codec = codec;
+    }
+
+    /**
+     * @param codec codec name
+     * @param blobNameFormat format of the blobname in {@link String#format} format
+     * @param reader         prototype object that can deserialize T from XContent
+     * @param wrapper        true if content should be surrounded with extra start/endObject during serialization
+     * @param compress      true if the content should be compressed
+     */
+    public ChecksumBlobStoreFormat(String codec, String blobNameFormat, FromXContentBuilder<T> reader, ParseFieldMatcher parseFieldMatcher, boolean wrapper, boolean compress) {
+        this(codec, blobNameFormat, reader, parseFieldMatcher, wrapper, compress, DEFAULT_X_CONTENT_TYPE);
+    }
+
+    public T readBlob(BlobContainer blobContainer, String blobName) throws IOException {
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            byte[] bytes = ByteStreams.toByteArray(inputStream);
+            final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+            try (ByteArrayIndexInput indexInput = new ByteArrayIndexInput(resourceDesc, bytes)) {
+                CodecUtil.checksumEntireFile(indexInput);
+                CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+                long filePointer = indexInput.getFilePointer();
+                long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+                BytesReference bytesReference = new BytesArray(bytes, (int)filePointer, (int)contentSize);
```

Comment:
That's not really possible. See my comments about byte arrays above. If file is too big, it would fail a few lines before when we try to read it:
```
byte[] bytes = ByteStreams.toByteArray(inputStream);
```